// vue/src/core/instance/state.js
/* @flow */

import config from '../config'
import Watcher from '../observer/watcher'
import Dep, { pushTarget, popTarget } from '../observer/dep'
import { isUpdatingChildComponent } from './lifecycle'

import {
  set,
  del,
  observe,
  defineReactive,
  toggleObserving
} from '../observer/index'

import {
  warn,
  bind,
  noop,
  hasOwn,
  hyphenate,
  isReserved,
  handleError,
  nativeWatch,
  validateProp,
  isPlainObject,
  isServerRendering,
  isReservedAttribute
} from '../util/index'

const sharedPropertyDefinition = {
  enumerable: true,
  configurable: true,
  get: noop,
  set: noop
}

export function proxy (target: Object, sourceKey: string, key: string) {
  sharedPropertyDefinition.get = function proxyGetter () {
    return this[sourceKey][key]
  }
  sharedPropertyDefinition.set = function proxySetter (val) {
    this[sourceKey][key] = val
  }
  Object.defineProperty(target, key, sharedPropertyDefinition)
}

export function initState (vm: Component) {
  vm._watchers = []
  const opts = vm.$options
    //处理 options.props 的成员
  if (opts.props) initProps(vm, opts.props)
    // 处理 options.methods 的成员
  if (opts.methods) initMethods(vm, opts.methods)

    // 处理 options.data 响应式化
  if (opts.data) {
    initData(vm)
  } else {
    observe(vm._data = {}, true /* asRootData */)
  }

  // 处理 options.computed 计算属性
  if (opts.computed) initComputed(vm, opts.computed)

    // 处理options.watch
  if (opts.watch && opts.watch !== nativeWatch) {
    initWatch(vm, opts.watch)
  }
}
/*
  主要分为三个阶段，
  第一个阶段，对每一个props做校验和求值
  第二阶段，求值后的props做成响应式
  第三个阶段，对props做一个代理
 */
function initProps (vm: Component, propsOptions: Object) {
  const propsData = vm.$options.propsData || {}  // 父组件传递给子组件的props数据
  const props = vm._props = {}  // 用来存储子组件计算后的props的值
  // cache prop keys so that future props updates can iterate using Array
  // instead of dynamic object key enumeration.
  const keys = vm.$options._propKeys = []  // props更新的时候用的
  const isRoot = !vm.$parent
  // root instance props should be converted
  if (!isRoot) {   // 响应式优化   非根实例   父组件已经对 data数据做了响应式的处理，这里不需要再对对象内的值做响应式处理， 其实父组件的某个值发生了变化，就会渲染整个组件，子组件也会重新渲染
    toggleObserving(false)
  }

  // 第一步，检验求值
  for (const key in propsOptions) {
    keys.push(key)
    const value = validateProp(key, propsOptions, propsData, vm)  // 对props做校验和求值
    /* istanbul ignore else */
    if (process.env.NODE_ENV !== 'production') {
      const hyphenatedKey = hyphenate(key)
      if (isReservedAttribute(hyphenatedKey) ||
          config.isReservedAttr(hyphenatedKey)) {
        warn(
          `"${hyphenatedKey}" is a reserved attribute and cannot be used as component prop.`,
          vm
        )
      }
      defineReactive(props, key, value, () => {
        if (!isRoot && !isUpdatingChildComponent) {
          warn(
            `Avoid mutating a prop directly since the value will be ` +
            `overwritten whenever the parent component re-renders. ` +
            `Instead, use a data or computed property based on the prop's ` +
            `value. Prop being mutated: "${key}"`,
            vm
          )
        }
      })
    } else {

      // 将属性 响应式化
      defineReactive(props, key, value)
    }
    // static props are already proxied on the component's prototype
    // during Vue.extend(). We only need to proxy props defined at
    // instantiation here.
    if (!(key in vm)) {
      proxy(vm, `_props`, key)   // 将_props上的成员映射到Vue的实例上，使得不需要app._props.xxx来访问，直接使用app.xxx来访问
    }
  }
  toggleObserving(true)   // 响应式设置完了之后，又把shouldObserve还原为初始值true
}

function initData (vm: Component) {
  let data = vm.$options.data

    // 将data挂载到 实例的_data上
  data = vm._data = typeof data === 'function'
    ? getData(data, vm)
    : data || {}
  if (!isPlainObject(data)) {
    data = {}
    process.env.NODE_ENV !== 'production' && warn(
      'data functions should return an object:\n' +
      'https://vuejs.org/v2/guide/components.html#data-Must-Be-a-Function',
      vm
    )
  }
  // proxy data on instance
  const keys = Object.keys(data)
  const props = vm.$options.props
  const methods = vm.$options.methods
  let i = keys.length
  while (i--) {
    const key = keys[i]

      // 这里判断是为了避免props,data,method等数据发生冲突，同名的问题
    if (process.env.NODE_ENV !== 'production') {
      if (methods && hasOwn(methods, key)) {
        warn(
          `Method "${key}" has already been defined as a data property.`,
          vm
        )
      }
    }
    if (props && hasOwn(props, key)) {
      process.env.NODE_ENV !== 'production' && warn(
        `The data property "${key}" is already declared as a prop. ` +
        `Use prop default value instead.`,
        vm
      )
    } else if (!isReserved(key)) {   // 如果 key不是以"_"或者"$"开头，  这两种开头的欧式Vue内部的属性或 这方法
      proxy(vm, `_data`, key)   //循环了data的所有属性，映射到Vue实例上，
                              //就不是必须使用 app._datga.xxx来访问了，
                              //而是使用app.xxx来访问了
    }
  }
  // observe data
  observe(data, true /* asRootData */)      //进行响应式化，
}

export function getData (data: Function, vm: Component): any {
  // #7573 disable dep collection when invoking data getters
  pushTarget()  // 由于此时 Vue的初始化，还没有进行模板渲染，所以不需要进行依赖收集，在pushTarget的时候传入空，就将全局的watcher 设置为undefined， 依赖收集的时候有一个判断，是Dep.target存在的时候才收集
  try {
    return data.call(vm, vm)
  } catch (e) {
    handleError(e, vm, `data()`)
    return {}
  } finally {
    popTarget()
  }
}

const computedWatcherOptions = { lazy: true }

function initComputed (vm: Component, computed: Object) {
  // $flow-disable-line
  const watchers = vm._computedWatchers = Object.create(null)
  // computed properties are just getters during SSR
  const isSSR = isServerRendering()

  for (const key in computed) {
    const userDef = computed[key]
    const getter = typeof userDef === 'function' ? userDef : userDef.get
    if (process.env.NODE_ENV !== 'production' && getter == null) {
      warn(
        `Getter is missing for computed property "${key}".`,
        vm
      )
    }

    if (!isSSR) {
      // create internal watcher for the computed property.
      watchers[key] = new Watcher(
        vm,   // vue的实例
        getter || noop,   // cmputed对象中属性的 get方法
        noop,
        computedWatcherOptions  //{ lazy: true }
      )
    }

    // component-defined computed properties are already defined on the
    // component prototype. We only need to define computed properties defined
    // at instantiation here.
    if (!(key in vm)) {   // 这个key在 vm 上海没有定义
      defineComputed(vm, key, userDef)
    } else if (process.env.NODE_ENV !== 'production') {
      //判断计算属性对于的 key 是否已经被 data 或者 prop 所占用，如果是的话则在开发环境报相应的警告。
      if (key in vm.$data) {
        warn(`The computed property "${key}" is already defined in data.`, vm)
      } else if (vm.$options.props && key in vm.$options.props) {
        warn(`The computed property "${key}" is already defined as a prop.`, vm)
      }
    }
  }
}

export function defineComputed (
  target: any,
  key: string,
  userDef: Object | Function   // computed 对象的属性所对应的属性值
) {
  const shouldCache = !isServerRendering()  // 在浏览器环境下 sholdCache 是true
  if (typeof userDef === 'function') {
    sharedPropertyDefinition.get = shouldCache
      ? createComputedGetter(key)
      : createGetterInvoker(userDef)
    sharedPropertyDefinition.set = noop
  } else {
    sharedPropertyDefinition.get = userDef.get
      ? shouldCache && userDef.cache !== false
        ? createComputedGetter(key)
        : createGetterInvoker(userDef.get)
      : noop
    sharedPropertyDefinition.set = userDef.set || noop
  }
  if (process.env.NODE_ENV !== 'production' &&
      sharedPropertyDefinition.set === noop) {
    sharedPropertyDefinition.set = function () {
      warn(
        `Computed property "${key}" was assigned to but it has no setter.`,
        this
      )
    }
  }
  Object.defineProperty(target, key, sharedPropertyDefinition)
}
// 浏览器触发的情况，里面会对数据的访问关联一个watcher
function createComputedGetter (key) {
  return function computedGetter () {
    const watcher = this._computedWatchers && this._computedWatchers[key]
    if (watcher) {
      if (watcher.dirty) {
        watcher.evaluate()   // 执行循序  watcher.evalute() -> watcher.get() ->  watcher.getter() -> 返回watcher.value
          /*
                             watcher.getter()  :方法 ，是 compute属性值中的 get方法（属性是函数的话就是当前函数，属性是对象的话，对应的就是对象中的 get方法
                              wathcer.value  :  这个值时 watcher.getter()函数的返回值（即 compute属性设置get函数的返回值）
                              watcher.dirty   :  watcher.dirty是用来控制缓存的，
                              1.第一次获取的时候，dirty为 true， 会调用一次 watcher.evaluate计算一次 watcher.value
                              调用后，在 watcher.evaluate()中 将dirty置为false，
                              2.再次获取 computed的对象的属性时， 由于 dirty为false, 会去之前的值，不会重新计算，
                              3.当 computed依赖的属性改变是，dirty会被置为 true, 
                       

           */
      }
      if (Dep.target) {
        watcher.depend()
      }
      return watcher.value
    }
  }
}

// 服务端渲染的时候触发，里面直接计算不会涉及watcher 处理
function createGetterInvoker(fn) {
  return function computedGetter () {
    return fn.call(this, this)
  }
}


function initMethods (vm: Component, methods: Object) {
  const props = vm.$options.props
  for (const key in methods) {
    if (process.env.NODE_ENV !== 'production') {
      if (typeof methods[key] !== 'function') {
        warn(
          `Method "${key}" has type "${typeof methods[key]}" in the component definition. ` +
          `Did you reference the function correctly?`,
          vm
        )
      }
      if (props && hasOwn(props, key)) {
        warn(
          `Method "${key}" has already been defined as a prop.`,
          vm
        )
      }
      if ((key in vm) && isReserved(key)) {
        warn(
          `Method "${key}" conflicts with an existing Vue instance method. ` +
          `Avoid defining component methods that start with _ or $.`
        )
      }
    }

    //将 methods属性中方法，绑定上下文后，挂载到vue实例上
    vm[key] = typeof methods[key] !== 'function' ? noop : bind(methods[key], vm)
  }
}

function initWatch (vm: Component, watch: Object) {
  for (const key in watch) {
    const handler = watch[key]
    if (Array.isArray(handler)) {
      for (let i = 0; i < handler.length; i++) {
        createWatcher(vm, key, handler[i])
      }
    } else {
      createWatcher(vm, key, handler)
    }
  }
}

function createWatcher (
  vm: Component,
  expOrFn: string | Function,   // key
  handler: any,     // 处理函数
  options?: Object
) {
  if (isPlainObject(handler)) {
    options = handler
    handler = handler.handler
  }
  if (typeof handler === 'string') {  // handler是一个methods方法名的时候
    handler = vm[handler]
  }
  return vm.$watch(expOrFn, handler, options)
}

export function stateMixin (Vue: Class<Component>) {
  // flow somehow has problems with directly declared definition object
  // when using Object.defineProperty, so we have to procedurally build up
  // the object here.
  const dataDef = {}
  dataDef.get = function () { return this._data }
  const propsDef = {}
  propsDef.get = function () { return this._props }
  if (process.env.NODE_ENV !== 'production') {
    dataDef.set = function () {
      warn(
        'Avoid replacing instance root $data. ' +
        'Use nested data properties instead.',
        this
      )
    }
    propsDef.set = function () {
      warn(`$props is readonly.`, this)
    }
  }
  Object.defineProperty(Vue.prototype, '$data', dataDef)
  Object.defineProperty(Vue.prototype, '$props', propsDef)

  Vue.prototype.$set = set
  Vue.prototype.$delete = del

  Vue.prototype.$watch = function (
    expOrFn: string | Function,
    cb: any,
    options?: Object
  ): Function {
    const vm: Component = this
    if (isPlainObject(cb)) {  // $watch 可以手动调用，来观测属性，所以 cb可能是一个对象
      return createWatcher(vm, expOrFn, cb, options)
    }
    options = options || {}
    options.user = true
    const watcher = new Watcher(vm, expOrFn, cb, options)
    if (options.immediate) {
      try {
        cb.call(vm, watcher.value)
      } catch (error) {
        handleError(error, vm, `callback for immediate watcher "${watcher.expression}"`)
      }
    }
    return function unwatchFn () {
      watcher.teardown()
    }
  }
}
