// vue/src/platforms/web/entry-runtime-with-compiler.js
/* @flow */

import config from 'core/config'
import { warn, cached } from 'core/util/index'
import { mark, measure } from 'core/util/perf'  //调试代码性能用的

import Vue from './runtime/index'  //  在这里定义了 一个mount方法   tag1
import { query } from './util/index'  //query 获取元素的方法

//  用来生成render的工具方法
import { compileToFunctions } from './compiler/index'
import { shouldDecodeNewlines, shouldDecodeNewlinesForHref } from './util/compat'

const idToTemplate = cached(id => {
  const el = query(id)
  return el && el.innerHTML
})

const mount = Vue.prototype.$mount  //(看上面tag1)获取mount方法 ，然后重新设置一个mount方法（编译版本的） ；runtime-only版本没有这个
//这个mount的作用就是解析 编译生产一个render函数，再有之间定义的mount处理
Vue.prototype.$mount = function (
  el?: string | Element,  //el 可以是个字符串 也可以是 element
  hydrating?: boolean
): Component {
  el = el && query(el)    //query获取dom对象

  /* istanbul ignore if */
  if (el === document.body || el === document.documentElement) {  //el如果是文档标签的话，就会报一个错误
    process.env.NODE_ENV !== 'production' && warn(   //会覆盖原理的html元素 和 body 元素，因为会将原来的元素覆盖掉
      `Do not mount Vue to <html> or <body> - mount to normal elements instead.`
    )
    return this
  }

  const options = this.$options
  // resolve template/el and convert to render function
  if (!options.render) {
    let template = options.template
    if (template) {
      if (typeof template === 'string') {
        if (template.charAt(0) === '#') {
          template = idToTemplate(template)
          /* istanbul ignore if */
          if (process.env.NODE_ENV !== 'production' && !template) {
            warn(
              `Template element not found or is empty: ${options.template}`,
              this
            )
          }
        }
      } else if (template.nodeType) {
        template = template.innerHTML
      } else {
        if (process.env.NODE_ENV !== 'production') {
          warn('invalid template option:' + template, this)
        }
        return this
      }
    } else if (el) {
      template = getOuterHTML(el)
    }
    if (template) {
      /* istanbul ignore if */
      if (process.env.NODE_ENV !== 'production' && config.performance && mark) {
        mark('compile')
      }

      const { render, staticRenderFns } = compileToFunctions(template, {
        outputSourceRange: process.env.NODE_ENV !== 'production',
        shouldDecodeNewlines,
        shouldDecodeNewlinesForHref,
        delimiters: options.delimiters,
        comments: options.comments
      }, this)
      options.render = render
      options.staticRenderFns = staticRenderFns

      /* istanbul ignore if */
      if (process.env.NODE_ENV !== 'production' && config.performance && mark) {
        mark('compile end')
        measure(`vue ${this._name} compile`, 'compile', 'compile end')
      }
    }
  }
  return mount.call(this, el, hydrating)
}

/**
 * Get outerHTML of elements, taking care
 * of SVG elements in IE as well.
 ie中对某些特定的标签是无法处理的
 获取外部的html标签
 */
function getOuterHTML (el: Element): string {
  if (el.outerHTML) {
    return el.outerHTML
  } else {
    //这里就是一个兼容性的语法
    const container = document.createElement('div')
    container.appendChild(el.cloneNode(true))
    return container.innerHTML
  }
}

Vue.compile = compileToFunctions

export default Vue
